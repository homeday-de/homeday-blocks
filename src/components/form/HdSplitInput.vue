<template>
  <div
    :class="fieldClasses"
    class="field field--double"
  >
    <img
      v-if="icon"
      :src="icon"
      role="presentation"
      class="field__icon"
    >
    <span class="field__border"/>
    <div class="field__double-input">
      <template
        v-for="(field, index) in fields">
        <input
          ref="inputs"
          :key="`split-input-${index}`"
          v-model="fieldsValues[field.name]"
          v-bind="field"
          :disabled="disabled"
          class="field__input"
          @focus="handleFocus(field)"
          @blur="handleBlur(field)">
        <div
          v-if="index < fields.length - 1"
          :key="`split-input-separator-${index}`"
          class="field__double-input__separator">
          <div
            class="field__double-input__separator__symbol"
            v-text="separator"/>
        </div>
      </template>
    </div>
    <label
      v-if="label"
      class="field__label"
    >
      {{ label }}
    </label>
    <p
      v-if="error"
      class="field__error"
    >
      {{ error }}
    </p>
    <p
      v-else-if="helper"
      class="field__error field__error--helper"
      v-html="helper"
    />
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';

export default {
  name: 'HdSplitInput',
  props: {
    label: {
      type: String,
      default: '',
    },
    fields: {
      type: Array,
      default: () => [],
      validator: fields => fields.every(({ name }) => {
        if (name === undefined || name === null) {
          // eslint-disable-next-line no-console
          console.warn('All fields must have a non-null `name` property');
          return false;
        }
        return true;
      }),
    },
    value: {
      type: Object,
      default: () => ({}),
    },
    separator: {
      type: String,
      default: 'â€“',
    },
    required: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
  },
  data() {
    return {
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
      fieldsValues: this.getFieldsValues(),
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isEmpty() {
      return Object.values(this.fieldsValues || {})
        .some(value => value === null || value === undefined || value === '');
    },
    fieldClasses() {
      return {
        'field--active': this.isActive,
        'field--filled': !this.isEmpty,
        'field--invalid': this.isValid === false,
        'field--disabled': this.disabled,
        'field--hasIcon': this.icon,
      };
    },
  },
  watch: {
    fieldsValues: {
      deep: true,
      handler(value) {
        this.validate();
        this.$emit('input', value);
      },
    },
  },
  mounted() {
    if (this.autofocus) {
      this.$refs.inputs[0].focus();
    }
  },
  methods: {
    handleFocus(field) {
      this.isActive = true;
      this.$emit('fieldFocus', field);
    },
    handleBlur(field) {
      this.isActive = false;
      this.validate();
      this.$emit('fieldBlur', field);
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    validate() {
      if (this.required && this.isEmpty) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else {
        this.hideError();
      }
      return !this.error;
    },
    getFieldsValues() {
      return this.fields.reduce((values, { name }) => ({
        ...values,
        [name]: this.value[name],
      }), {});
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

.field {
  $f: &;
  &__double-input {
    &::before, &::after {
      display: none;
    }
    &__separator {
      display: flex;
      align-items: center;
      justify-content: center;
      flex: 0 0 $inline-m;
      z-index: 1;
      background-color: $inputBackground;
      color: getShade($quaternary-color, 80);
      font-size: 20px;
      font-weight: 700;

      &__symbol {
        margin-top: $stack-m;
      }
    }
  }
  &__error {
    width: 100%;
    text-align: left;
    &--helper {
      display: block;
      color: getShade($quaternary-color, 80);
    }
  }
}
.field--double .field__input {
  width: 100%;
}
</style>
