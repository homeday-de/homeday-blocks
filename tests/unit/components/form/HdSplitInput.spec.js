import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import FIELD_CLASSES from './FIELD_CLASSES';
import HdSplitInput from '@/components/form/HdSplitInput.vue';

const ERROR_SELECTOR = '.field__error';
const HELPER_SELECTOR = '.field__error--helper';
const SEPARATOR_SELECTOR = '.field__double-input__separator__symbol';
const ICON_SELECTOR = '.field__icon';

const TEST_VALUE = 'new value';
const TEST_FIELDS = [
  {
    placeholder: 'Input 1',
    name: 'input1',
  },
  {
    placeholder: 'Input 2',
    name: 'input2',
  },
];

const wrapperBuilder = wrapperFactoryBuilder(HdSplitInput, {
  props: {
    name: 'test name',
    label: 'test label',
    fields: TEST_FIELDS,
  },
});

describe('HdSplitInput', () => {
  it('is rendered as expected', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('is rendered with prefilled value', () => {
    const TEST_NAME = TEST_FIELDS[0].name;
    const wrapper = wrapperBuilder({
      props: {
        value: {
          [TEST_NAME]: TEST_VALUE,
        },
      },
    });

    expect(wrapper.find(`[name=${TEST_NAME}]`).element.value).toBe(TEST_VALUE);
  });

  it('can customize the separator', () => {
    const CUSTOM_SEPARATOR = '777';
    const wrapper = wrapperBuilder({
      props: {
        separator: CUSTOM_SEPARATOR,
      },
    });
    const wrongSeparators = wrapper
      .findAll(SEPARATOR_SELECTOR)
      .filter(separator => separator.text() !== CUSTOM_SEPARATOR);

    expect(wrongSeparators.length).toBe(0);
  });

  describe('emits events with the right payload', () => {
    it('input', async () => {
      const wrapper = wrapperBuilder();
      const TEST_NAME = TEST_FIELDS[0].name;

      wrapper.find(`[name=${TEST_NAME}]`).setValue(TEST_VALUE);

      await wrapper.vm.$nextTick();

      expect(wrapper.emitted('input')[0][0][TEST_NAME]).toBe(TEST_VALUE);
    });

    it('fieldFocus', () => {
      const wrapper = wrapperBuilder();
      const TEST_FIELD = TEST_FIELDS[0];

      wrapper.find(`[name=${TEST_FIELD.name}]`).trigger('focus');

      expect(wrapper.emitted('fieldFocus')[0][0]).toEqual(TEST_FIELD);
    });

    it('fieldBlur', () => {
      const wrapper = wrapperBuilder();
      const TEST_FIELD = TEST_FIELDS[1];

      wrapper.find(`[name=${TEST_FIELD.name}]`).trigger('blur');

      expect(wrapper.emitted('fieldBlur')[0][0]).toEqual(TEST_FIELD);
    });
  });

  it('allows setting a helper', async () => {
    const wrapper = wrapperBuilder();
    const TEST_HTML = '<b>test</b> test';

    wrapper.vm.showHelper(TEST_HTML);

    await wrapper.vm.$nextTick();

    expect(wrapper.find(HELPER_SELECTOR).element.innerHTML).toBe(TEST_HTML);
  });

  it('validates requiredness', async () => {
    const wrapper = wrapperBuilder({
      props: {
        value: {},
        required: true,
      },
    });

    expect(wrapper.vm.validate()).toBe(false);

    await wrapper.vm.$nextTick();

    expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
    expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);

    wrapper.setProps({
      required: false,
    });

    expect(wrapper.vm.validate()).toBe(true);
  });

  it('Supports disabling', () => {
    const wrapper = wrapperBuilder({
      props: {
        disabled: true,
      },
    });
    const enabledInputs = wrapper
      .findAll('input')
      .filter(input => input.attributes().disabled !== 'disabled');

    expect(enabledInputs.length).toBe(0);
  });

  it('Can render an icon', async () => {
    const wrapper = wrapperBuilder();
    const ICON_PATH = 'fake/icon.svg';

    expect(wrapper.classes()).not.toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(false);

    wrapper.setProps({
      icon: ICON_PATH,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.classes()).toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(true);
  });
});
