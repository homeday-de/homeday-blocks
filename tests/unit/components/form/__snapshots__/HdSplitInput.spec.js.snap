// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdSplitInput is rendered as expected 1`] = `
<div class="field field--minimized-label" name="test name" label="test label">
  <!---->
  <div class="field__body">
    <div class="field__main">
      <div class="split-input"><input placeholder="Input 1" name="input1" class="split-input__input">
        <div class="split-input__separator">
          <div class="split-input__separator-symbol">–</div>
        </div><input placeholder="Input 2" name="input2" class="split-input__input">
        <!---->
      </div> <label for="input1" class="field__label">
        test label
      </label>
      <div class="field__input-right"></div>
      <div class="field__border"></div>
    </div>
    <p class="field__helper">‍</p>
  </div>
</div>
`;
